core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1081,8 +1094,18 @@ public Builder setCurrentSchema(int schemaId) {
       return this;
     }
 
+    public Builder addSchema(Schema schema) {
+      addSchemaInternal(schema, schema.highestFieldId());
```

Comment:
I think this needs to be `Math.max(lastColumnId, schema.highestFieldId())` because the new schema could drop a column.

Reply:
Great catch! Thanks!